---
title: "Introduction to Bioinformatics"
output: html_document
---

load libraries and define functions
```{r}
library(DESeq2) 
library(limma)
library(EnhancedVolcano)
library(AnnotationDbi)
library(dplyr)
library(msigdbr)
library(clusterProfiler)
library(scales)
library(ggpubr)
library(ggplot2)
library(enrichplot)

# function to read experiment designs from our datasets
read_design <- function(filename) {
  E <- read.csv(filename, sep = "\t")

  rownames(E) <- E$Run
  
  E <- E[c("Sample.Characteristic.disease.", "Sample.Characteristic.organism.part.")]
  colnames(E) <- c("Diagnosis", "Brodmann")

  E[E == "bipolar disorder"] = "bipolar"
  E$`Brodmann` <- sub("Brodmann \\(1909\\) area ", "", E$`Brodmann`)
  
  E[E == "dorsolateral prefrontal cortex"] = "46"
  
  return(E)
}

```

Not Used
```{}
# for unused data
read_processed_data <- function(filename){
  data <- read.delim(filename, sep="\t")
  data <- cbind(data[,1], data)
  data <- data[-1,]

  filtering <- select(hgu133a.db,keys= data$Scan.REF, columns=c("PROBEID", "ENSEMBL", "SYMBOL"))

  filtering <- filtering[!duplicated(filtering$ENSEMBL),]
  filtering <- filtering[!duplicated(filtering$PROBEID),]
  filtering <- filtering[!is.na(filtering$ENSEMBL),]

  data <- data[data$Scan.REF %in% filtering$PROBEID,]
  data[1:2] <- filtering[2:3]
  colnames(data)[1:2] <- c("Gene.ID", "Gene Name")
  
  return(data)
}

data3 <- read_processed_data("../Databases/Bipolar - E-GEOD-5388/E-GEOD-5388-processed-data-1435128605.txt")

rownames(data3) <- data3[,1]

metadata <- read.delim("../Databases/Bipolar - E-GEOD-5388/E-GEOD-5388-experiment-design.tsv")[1:3][-2]

colnames(metadata) <- c("Speciman", "Diagnosis")

rownames(metadata) <- metadata[,1]
metadata <- metadata[-1]

counts <- data3
counts <- counts[-1][-1]
```

Load Data
```{r}
# read the experiments' designs
E7 <- read_design("../Databases/Bipolar - E-GEOD-78936/E-GEOD-78936-experiment-design.tsv")
E5 <- read_design("../Databases/Bipolar - E-GEOD-53239/E-GEOD-53239-experiment-design.tsv")

# add identifiers for each dataset
E5 <- cbind(rep("GEOD-53239", times=nrow(E5)), E5)
E7 <- cbind(rep("GEOD-78936", times=nrow(E7)), E7)

colnames(E5)[1] = colnames(E7)[1] = "Dataset"

# combine metadata
metadata <- rbind(E5, E7)

# read counts data
data1 <- read.delim("../Databases/Bipolar - E-GEOD-53239/E-GEOD-53239-raw-counts.tsv")
data2 <- read.delim("../Databases/Bipolar - E-GEOD-78936/E-GEOD-78936-raw-counts.tsv")

# combine only the counts data
counts <- cbind(data1[-1][-1], data2[-1][-1])
counts <- counts[sort(colnames(counts))]

# take the gene names from the counts data
genes.symbols <- data1[,2]
```


Run DESeq2
```{r}
# run DESeq2 on combined data from both experiments
# using Dataset to distinguish between them which will remove the batch effect. 
# Since Dataset is a division contained in Brodmann, using Brodmann is enough
dds <- DESeqDataSetFromMatrix(countData=counts, colData=metadata, design= ~Brodmann + Diagnosis)
dds <- DESeq(dds)

# run different DESeq2 for each brain area
areas <- c(9, 11, 24, 46)
areaDds <- list()
i <- 1
for (i in seq_along(areas)){
  areaDds[[i]] <- DESeqDataSetFromMatrix(countData=counts[,metadata$Brodmann == areas[i]], colData=metadata[metadata$Brodmann == areas[i],], design= ~Diagnosis)
  areaDds[[i]] <- DESeq(areaDds[[i]])
}
```


Analyze the data:
```{r}
# do logfold change for Diagnosis
# normal vs bipolar
resLFC.bipolar.normal <- lfcShrink(dds, contrast = c("Diagnosis", "bipolar", "normal"), type = "ashr")
resLFC.bipolar.normal["symbol"] <- genes.symbols
resOrdered.bipolar.normal <- resLFC.bipolar.normal[order(resLFC.bipolar.normal$pvalue),]

# bipolar vs schizophrenia
resLFC.bipolar.schizo <- lfcShrink(dds, contrast = c("Diagnosis", "bipolar", "schizophrenia"), type = "ashr")
resLFC.bipolar.schizo["symbol"] <- genes.symbols
resOrdered.bipolar.schizo <- resLFC.bipolar.schizo[order(resLFC.bipolar.schizo$pvalue),]

# differentiate brain areas
resLFC.areas <- list()
resOrdered.areas <- list()
comparisons <- list()
for (i in seq_along(areas)){
  controls = c("normal", "schizophrenia")
  if (areas[i] == 46){
    controls = controls[1]
  }
  
  for (control in controls){
    comparison <- paste0("bipolar vs ", control, " in ", areas[i])
    print(paste0("Calculating ", comparison))
    resLFC.areas[[i]] <- lfcShrink(areaDds[[i]], contrast = c("Diagnosis", "bipolar", control), type = "ashr")
    resLFC.areas[[i]]["symbol"] <- genes.symbols
    resOrdered.areas[[i]] <- resLFC.areas[[i]][order(resLFC.areas[[i]]$pvalue),]
    comparisons[[i]] <- comparison
  }
}


# normalize our data
dds.norm <- vst(dds, blind=F)
rownames(dds.norm) <- genes.symbols

# remove batch effect for PCA
mat <- assay(dds.norm)
mat <- removeBatchEffect(mat, dds.norm$Dataset)
assay(dds.norm) <- mat
```


Make some nice looking graphs :)
```{r}
# create a volcano plot
EnhancedVolcano(resOrdered.bipolar.normal, lab = resOrdered.bipolar.normal$symbol, x = 'log2FoldChange', y = 'padj', FCcutoff=2, pCutoff = 10e-10, title="bipolar vs normal")
EnhancedVolcano(resOrdered.bipolar.schizo, lab = resOrdered.bipolar.schizo$symbol, x = 'log2FoldChange', y = 'padj', FCcutoff=2, pCutoff = 10e-10, title="bipolar vs schizophrenia")
for (i in seq_along(comparisons)){
  print(EnhancedVolcano(resOrdered.areas[[i]], lab = resOrdered.areas[[i]]$symbol, x = 'log2FoldChange', y = 'padj', FCcutoff=2, pCutoff = 10e-10, title=comparisons[i]))
}

# make some pca plots
plotPCA(dds.norm, intgroup=c("Diagnosis"))
plotPCA(dds.norm, intgroup=c("Brodmann"))
plotPCA(dds.norm, intgroup=c("Diagnosis", "Brodmann"))
```

Some count plots for significant genes :)
```{r}
rownames(dds) <- genes.symbols

genes <- c("LINC02340", "MTND6P4", "MT1X", "IL1RL1")

means_comp <- list(c("bipolar", "normal"), c("bipolar", "schizophrenia"), c("normal", "schizophrenia"))

for (gene in genes) {
  gene.index <- which(genes.symbols %in% gene)
  plot <- plotCounts(dds, gene = rownames(dds)[gene.index], intgroup = c('Diagnosis'), xlab="Diagnosis", title=paste0(gene, " counts"), returnData = T)
  print(
    ggplot(plot, aes(x = Diagnosis, y = count, fill=Diagnosis)) + 
      geom_violin(draw_quantiles = 0.5) + 
      geom_jitter(color="royalblue", size=0.6, alpha=0.5) + 
      theme(legend.position="none")  + 
      ggtitle(paste0("Gene Expression of ", gene)) +
      scale_y_continuous(trans = log10_trans()) + 
      stat_compare_means(label = "p.signif", comparisons = means_comp)
    )
}

areaSpecific <- data.frame(GeneName = c("IGKC", "CHI3L2", "MT1X", "MTND6P4"), Brodmann = c(46, 46, 46, 46))

for (geneIndex in 1:nrow(areaSpecific)) {
  area <- areaSpecific[geneIndex,]$Brodmann
  gene <- areaSpecific[geneIndex,]$GeneName
  thisDds = areaDds[[which(areas == area)]]
  rownames(thisDds) <- genes.symbols
  plot <- plotCounts(thisDds, gene = gene, intgroup = c('Diagnosis'), xlab="Diagnosis", title=paste0(gene, " counts"), returnData = T)
  
  comp = means_comp
  if (area == 46){
    comp = means_comp[1]
  }
  
  print(
    ggplot(plot, aes(x = Diagnosis, y = count, fill=Diagnosis)) + 
      scale_fill_manual(values=c("#f8766d", "#00ba38", "#619cff")) +
      geom_violin(draw_quantiles = 0.5) + 
      geom_jitter(color="royalblue", size=0.6, alpha=0.5) + 
      theme(legend.position="none")  + 
      ggtitle(paste0("Gene Expression of ", gene, " in ", area)) + 
      scale_y_continuous(trans = log10_trans()) + 
      stat_compare_means(label = "p.signif", comparisons = comp)
    )
}
```

And Pathway analysis :)
```{r}
# try subtyping and checking them here

# remove NAs and create sorted genes list with LFC values and ENSMBLE ID as names
DE_genes_entrez_rank <- resOrdered.bipolar.normal[!is.na(resOrdered.bipolar.normal$padj),]
DE_genes_list <- DE_genes_entrez_rank$log2FoldChange
names(DE_genes_list) <- DE_genes_entrez_rank$symbol
DE_genes_list <- sort(DE_genes_list, decreasing=TRUE)

# collect hallmarks
hallmarks <- msigdbr(species = "Homo sapiens", category = "H") %>% dplyr::select(gs_name, gene_symbol)

# set nPerm to million so it's absolutely precise and does not change when the random seed changes
hm <- GSEA(DE_genes_list, TERM2GENE=hallmarks, nPerm=1000000)

# create a ridgeplot
ridgeplot(hm) + labs(x = "enrichment distribution")

# and a dot plot
dotplot(hm, showCategory = Inf, title="Pathway Analisys with GSEA, normal VS bipolar")

# We tried running the same code on resOrdered.bipolar.schizo, but it found no enriched terms
```
Clustering (not checked yet)
```{r}
normcounts <- assay(vst(dds[,dds$Diagnosis == "bipolar"], blind=T))
colnames(normcounts) <- paste0("bipolar", seq(1, ncol(normcounts)))
var_per_gene <- apply(normcounts, 1, var)
selectedGenes <- names(var_per_gene[order(var_per_gene, decreasing = T)][1:29])
normcounts.top1Kvar <- t(normcounts[selectedGenes,])
dist_mat <- dist(normcounts.top1Kvar, method = 'euclidean')

hclust_avg <- hclust(dist_mat, method = 'average')
plot(hclust_avg, cex = 0.6, hang = -1)
rect.hclust(hclust_avg, k = 2, border = 2:8)
```

Heatmap cause we can :(
```{r}
library(pheatmap)
library(org.Hs.eg.db)

normcounts <- assay(vst(dds[,dds$Diagnosis != "schizophrenia"], blind=T))

dds.symbol <- dds
rownames(dds.symbol) <- mapIds(org.Hs.eg.db,
                              keys=data1$Gene.ID,
                              column="SYMBOL",
                              keytype="ENSEMBL",
                              multiVals="first")

rownames(dds.symbol)[is.na(rownames(dds.symbol))] <- rownames(dds)[is.na(rownames(dds.symbol))]
rownames(dds.symbol) <- make.unique(rownames(dds.symbol))

df <- data.frame(row.names = colnames(dds.symbol),
                 diagnosis = colData(dds.symbol)$Diagnosis)

# Take top 10 genes with the lowest p-value that express in Bipolar (log2FoldChange>0)
selectUp <- resOrdered.bipolar.normal$symbol[resOrdered.bipolar.normal$log2FoldChange>0][1:20]
# Take top 10 genes with the lowest p-value that express in Healthy (log2FoldChange<0)
selectDown <- resOrdered.bipolar.normal$symbol[resOrdered.bipolar.normal$log2FoldChange<0][1:20]

selected <- c(selectUp,selectDown)

pheatmap(normcounts[selected,], cluster_rows=TRUE,
         show_colnames = FALSE, cluster_cols=TRUE, 
         annotation_col=df, scale = 'row', cutree_cols = 2, cutree_rows = 2)
```

