---
title: "Introduction to Bioinformatics"
output: html_document
---

# Setup

load libraries and define functions
```{r setup}
library(DESeq2) 
library(limma)
library(EnhancedVolcano)
library(AnnotationDbi)
library(dplyr)
library(msigdbr)
library(clusterProfiler)
library(scales)
library(ggpubr)
library(ggplot2)
library(enrichplot)
library(factoextra)
library(pheatmap)
library(xCell)
library(dendextend)

# function to read experiment designs from our datasets
read_design <- function(filename) {
  E <- read.csv(filename, sep = "\t")

  rownames(E) <- E$Run
  
  E <- E[c("Sample.Characteristic.disease.", "Sample.Characteristic.organism.part.")]
  colnames(E) <- c("Diagnosis", "Brodmann")

  E[E == "bipolar disorder"] = "bipolar"
  E$`Brodmann` <- sub("Brodmann \\(1909\\) area ", "", E$`Brodmann`)
  
  E[E == "dorsolateral prefrontal cortex"] = "46"
  
  return(E)
}

```

Load Data
```{r data-loading}
# read the experiments' designs
E7 <- read_design("../Databases/Bipolar - E-GEOD-78936/E-GEOD-78936-experiment-design.tsv")
E5 <- read_design("../Databases/Bipolar - E-GEOD-53239/E-GEOD-53239-experiment-design.tsv")

# add identifiers for each dataset
E5 <- cbind(rep("GEOD-53239", times=nrow(E5)), E5)
E7 <- cbind(rep("GEOD-78936", times=nrow(E7)), E7)

colnames(E5)[1] = colnames(E7)[1] = "Dataset"

# combine metadata
metadata <- rbind(E5, E7)

# read counts data
data1 <- read.delim("../Databases/Bipolar - E-GEOD-53239/E-GEOD-53239-raw-counts.tsv")
data2 <- read.delim("../Databases/Bipolar - E-GEOD-78936/E-GEOD-78936-raw-counts.tsv")

# combine only the counts data
counts <- cbind(data1[-1][-1], data2[-1][-1])
counts <- counts[sort(colnames(counts))]

# take the gene names from the counts data
genes.symbols <- data1[,2]

# if there is no gene name, use the ID
genes.symbols[genes.symbols == ""] <- data1$Gene.ID[genes.symbols == ""]

# replace dupliactes with ID too
genes.symbols[duplicated(genes.symbols)] <- data1$Gene.ID[duplicated(genes.symbols)]
```

# Identifying Biomarker Genes

Run DESeq2
```{r DESeq}
# run DESeq2 on combined data from both experiments
# using Dataset to distinguish between them which will remove the batch effect. 
# Since Dataset is a division contained in Brodmann, using Brodmann is enough
dds <- DESeqDataSetFromMatrix(countData=counts, colData=metadata, design= ~Brodmann + Diagnosis)
dds <- DESeq(dds)

rownames(dds) <- genes.symbols

# run different DESeq2 for each brain area
areas <- c(9, 11, 24, 46)
areaDds <- list()

for (i in seq_along(areas)){
  areaDds[[i]] <- DESeqDataSetFromMatrix(countData=counts[,metadata$Brodmann == areas[i]], colData=metadata[metadata$Brodmann == areas[i],], design= ~Diagnosis)
  areaDds[[i]] <- DESeq(areaDds[[i]])
}
```


Analyze the data:
```{r DESeq-analysis}
# do logfold change for Diagnosis
# normal vs bipolar
resLFC.bipolar.normal <- lfcShrink(dds, contrast = c("Diagnosis", "bipolar", "normal"), type = "ashr")
resLFC.bipolar.normal["symbol"] <- genes.symbols
resOrdered.bipolar.normal <- resLFC.bipolar.normal[order(resLFC.bipolar.normal$pvalue),]

# bipolar vs schizophrenia
resLFC.bipolar.schizo <- lfcShrink(dds, contrast = c("Diagnosis", "bipolar", "schizophrenia"), type = "ashr")
resLFC.bipolar.schizo["symbol"] <- genes.symbols
resOrdered.bipolar.schizo <- resLFC.bipolar.schizo[order(resLFC.bipolar.schizo$pvalue),]

# differentiate brain areas
resLFC.areas <- list()
resOrdered.areas <- list()
comparisons <- list()

j <- 1

for (i in seq_along(areas)){
  controls = c("normal", "schizophrenia")
  if (areas[i] == 46){
    # area 46 has no SZ patients
    controls = controls[1]
  }
  
  for (control in controls){
    comparison <- paste0("bipolar vs ", control, " in ", areas[i])
    print(paste0("Calculating ", comparison))
    resLFC.areas[[j]] <- lfcShrink(areaDds[[i]], contrast = c("Diagnosis", "bipolar", control), type = "ashr")
    resLFC.areas[[j]]["symbol"] <- genes.symbols
    resOrdered.areas[[j]] <- resLFC.areas[[j]][order(resLFC.areas[[j]]$pvalue),]
    comparisons[[j]] <- comparison
    j = j + 1
  }
}


# normalize our data
dds.norm <- vst(dds, blind=F)
rownames(dds.norm) <- genes.symbols

# remove batch effect for PCA
mat <- assay(dds.norm)
mat <- removeBatchEffect(mat, dds.norm$Dataset)
assay(dds.norm) <- mat
```


Make some nice looking graphs :)
```{r volcano-and-pca-plots}
# create volcano plots
EnhancedVolcano(resOrdered.bipolar.normal, lab = resOrdered.bipolar.normal$symbol, x = 'log2FoldChange', y = 'padj', FCcutoff=2, pCutoff = 10e-10, title="bipolar vs normal")
EnhancedVolcano(resOrdered.bipolar.schizo, lab = resOrdered.bipolar.schizo$symbol, x = 'log2FoldChange', y = 'padj', FCcutoff=2, pCutoff = 10e-10, title="bipolar vs schizophrenia")
for (i in seq_along(comparisons)){
  print(EnhancedVolcano(resOrdered.areas[[i]], lab = resOrdered.areas[[i]]$symbol, x = 'log2FoldChange', y = 'padj', FCcutoff=2, pCutoff = 10e-10, title=comparisons[i]))
}

# make some pca plots
plotPCA(dds.norm, intgroup=c("Diagnosis"))
plotPCA(dds.norm, intgroup=c("Brodmann"))
plotPCA(dds.norm, intgroup=c("Diagnosis", "Brodmann"))
```

Some count plots for significant genes :)
```{r count-plots}
# significant genes to check in violin plots
genes <- c("LINC02340", "MTND6P4", "MT1X", "IL1RL1")

# compare the means of each two groups
means_comp <- list(c("bipolar", "normal"), c("bipolar", "schizophrenia"), c("normal", "schizophrenia"))

# make the violin plots
for (gene in genes) {
  gene.index <- which(genes.symbols %in% gene)
  plot <- plotCounts(dds, gene = rownames(dds)[gene.index], intgroup = c('Diagnosis'), xlab="Diagnosis", title=paste0(gene, " counts"), returnData = T)
  print(
    ggplot(plot, aes(x = Diagnosis, y = count, fill=Diagnosis)) + 
      geom_violin(draw_quantiles = 0.5) + 
      geom_jitter(color="royalblue", size=0.6, alpha=0.5) + 
      theme(legend.position="none")  + 
      ggtitle(paste0("Gene Expression of ", gene)) +
      scale_y_continuous(trans = log10_trans()) + 
      stat_compare_means(label = "p.signif", comparisons = means_comp,  method = "wilcox.test")
    )
}

# significant genes in single areas to check in violin plots
areaSpecific <- data.frame(GeneName = c("IGKC", "CHI3L2", "MT1X", "MTND6P4"), Brodmann = c(46, 46, 46, 46))

# make the violin plots
for (geneIndex in 1:nrow(areaSpecific)) {
  area <- areaSpecific[geneIndex,]$Brodmann
  gene <- areaSpecific[geneIndex,]$GeneName
  thisDds = areaDds[[which(areas == area)]]
  rownames(thisDds) <- genes.symbols
  plot <- plotCounts(thisDds, gene = gene, intgroup = c('Diagnosis'), xlab="Diagnosis", title=paste0(gene, " counts"), returnData = T)
  
  comp = means_comp
  if (area == 46){
    # area 46 has no SZ patients
    comp = means_comp[1]
  }
  
  print(
    ggplot(plot, aes(x = Diagnosis, y = count, fill=Diagnosis)) + 
      scale_fill_manual(values=c("#f8766d", "#00ba38", "#619cff")) +
      geom_violin(draw_quantiles = 0.5) + 
      geom_jitter(color="royalblue", size=0.6, alpha=0.5) + 
      theme(legend.position="none")  + 
      ggtitle(paste0("Gene Expression of ", gene, " in ", area)) + 
      scale_y_continuous(trans = log10_trans()) + 
      stat_compare_means(label = "p.signif", comparisons = comp, method = "wilcox.test")
    )
}
```

# Identifying Enriched Pathways

And Pathway analysis :)
```{r pathway-analysis}
# remove NAs and create sorted genes list with LFC values and ENSMBLE ID as names
DE_genes_entrez_rank <- resOrdered.bipolar.normal[!is.na(resOrdered.bipolar.normal$padj),]
DE_genes_list <- DE_genes_entrez_rank$log2FoldChange
names(DE_genes_list) <- DE_genes_entrez_rank$symbol
DE_genes_list <- sort(DE_genes_list, decreasing=TRUE)

# collect hallmarks
hallmarks <- msigdbr(species = "Homo sapiens", category = "H") %>% dplyr::select(gs_name, gene_symbol)

# set nPerm to million so it's absolutely precise and does not change when the random seed changes
hm <- GSEA(DE_genes_list, TERM2GENE=hallmarks, nPerm=1000000)

# create a ridgeplot
ridgeplot(hm) + labs(x = "enrichment distribution")

# and a dot plot
dotplot(hm, showCategory = 178, title="Pathway Analisys with GSEA, normal VS bipolar")

# We tried running the same code on resOrdered.bipolar.schizo, but it found no enriched terms
```

# Examining the Changes in the Cellular Composition

```{r run-xCell}
# set rownames of counts so xCell understands what are the genes
rownames(counts) <- genes.symbols

# run the analysis
counts.xCell <- xCellAnalysis(counts, rnaseq=T)

p <- pheatmap(counts.xCell, show_colnames = F, silent = T)

# filter results as most of it is not valuble information
counts.xCell.filtered <- counts.xCell[rowMeans(counts.xCell) > 0.01,]

# scale for heatmap, avarage should be 0
counts.xCell.filtered.scaled <- t(scale(t(counts.xCell.filtered)))

# clip values
counts.xCell.filtered.scaled.clipped <- counts.xCell.filtered.scaled
counts.xCell.filtered.scaled.clipped[counts.xCell.filtered.scaled > 1] <- 1
counts.xCell.filtered.scaled.clipped[counts.xCell.filtered.scaled < -1] <- -1
```

```{r show results}
# clustering metadata for heatmap
df <- data.frame(row.names = colnames(dds),
                 diagnosis = colData(dds)$Diagnosis,
                 area = colData(dds)$Brodmann)

# show results
pheatmap(counts.xCell.filtered.scaled.clipped, cluster_rows=TRUE,
         show_colnames = FALSE, cluster_cols=TRUE, 
         annotation_col=df, scale = 'row', cutree_cols = 4, cutree_rows = 2)

# analyze clustering
k = 2
myclusters <- cutree(p$tree_col, k=k)

# show correlation between clusters and known factors about the patients
patients <- table(metadata$Diagnosis, myclusters)
mosaicplot(patients)

patients <- table(metadata$Brodmann, myclusters)
mosaicplot(patients)

# create every combination of clusters to see difference in the data
comps <- t(combn(1:k, 2))
comps <- split(comps, seq(nrow(comps)))

# violin plots of cell score in every cluster
for (cellType in rownames(counts.xCell.filtered)) {
  MEscore.df <- data.frame("score" = counts.xCell[cellType,],
                         "cluster" = factor(myclusters),
                         row.names = names(myclusters))
  
  print(
    ggplot(MEscore.df, aes(x=cluster, y=score, fill=cluster)) +
    geom_violin(draw_quantiles = 0.5) +
    labs(x="", y="Score") + stat_compare_means(method = "wilcox.test", comparisons = comps, label = "p.signif") +
      ggtitle(paste0("Cluster comparison of ", cellType))
  )
}

# see how the cell type divide the groups on a plot
chosenCells = c("Th1 cells", "Smooth muscle", "MSC")
scores.df <- data.frame("cluster" = factor(myclusters),
                         row.names = names(myclusters),
                        area=paste0(metadata$Brodmann, ":", factor(myclusters)))

# create a 
for (cell in chosenCells) {
  scores.df[cell] <- counts.xCell[cell,]
}

# create a list of every two cell type combination from choseCells
comps <- t(combn(1:length(chosenCells), 2))
comps <- split(comps, seq(nrow(comps)))

# see how well each two cell types divide the patients to the clusters
for (comp in comps){
  print(
    ggplot(scores.df, aes(x=scores.df[,chosenCells[comp[1]]], y=scores.df[,chosenCells[comp[2]]], color=cluster)) + 
      geom_point() + 
      xlab(chosenCells[comp[1]]) + 
      ylab(chosenCells[comp[2]])
  )
}
```

Overall, we can see that the division is quite good, and that there is some kind of difference between the clusters, but nothing that is expressed in other medical data we have

# Classifying Patients

Clustering
```{r clustering}
# get normalized counts only for bipolar
sampleFilter <- dds$Diagnosis != "schizophrenia"
normcounts <- assay(vst(dds[,sampleFilter], blind=T))
colnames(normcounts) <- paste0(dds$Diagnosis[sampleFilter], "_", seq(1, ncol(normcounts)))
var_per_gene <- apply(normcounts, 1, var)

# take 1000 most significant genes
selectedGenes <- names(var_per_gene[order(var_per_gene, decreasing = T)][1:1000])
selectedGenes <- genes

normcounts.topGenes <- t(normcounts[selectedGenes,])

# make an elbow plot (optimal k seems to be 2 or 3)
fviz_nbclust(normcounts.topGenes, FUN = hcut, method = "wss")

dist_mat <- dist(normcounts.topGenes, method = 'euclidean')

# plot the cluster dendrogram
hclust_avg <- dist_mat %>% hclust(method = 'average') %>% as.dendrogram

# set colors
colorlist <- labels(hclust_avg)
colorlist <- sub("normal.*", "#00ba38", colorlist)
colorlist <- sub("bipolar.*", "#f8766d", colorlist)
colorlist <- sub("schizophrenia.*", "#619cff", colorlist)

hclust_avg <- hclust_avg %>% 
  set("leaves_pch", 19) %>%
  set("leaves_cex", 1) %>%
  set("leaves_col", colorlist) %>%
  set("labels", "")

plot(hclust_avg, ex = 0.6, hang = -1, xlab="", ylab ="", sub="", axes=F)
rect.dendrogram(hclust_avg, k = 3, border = 2:8)
```

Heatmap
```{r heatmap}
# take normalized counts for normal and bipolar
sampleFilter <- dds$Diagnosis != "schizophrenia"
normcounts <- assay(vst(dds[,sampleFilter], blind=T))

# create labels for heatmap
df <- data.frame(row.names = colnames(dds[,sampleFilter]),
                 diagnosis = colData(dds[,sampleFilter])$Diagnosis,
                  area = colData(dds[,sampleFilter])$Brodmann)

# Take top 10 genes with the lowest p-value that express in Bipolar (log2FoldChange>0)
selectUp <- resOrdered.bipolar.normal$symbol[resOrdered.bipolar.normal$log2FoldChange>0][1:10]
# Take top 10 genes with the lowest p-value that express in Healthy (log2FoldChange<0)
selectDown <- resOrdered.bipolar.normal$symbol[resOrdered.bipolar.normal$log2FoldChange<0][1:10]

selected <- c(selectUp,selectDown)

# create the heatmap
pheatmap(normcounts[selected,], cluster_rows=TRUE,
         show_colnames = FALSE, cluster_cols=TRUE, 
         annotation_col=df, scale = 'row', cutree_cols = 2, cutree_rows = 2)
```